### PR TITLE
[DA][15/n][asset-graph-entity] Create single base class for AssetKey and AssetCheckKey

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

The start of a larger stack of changes to make AssetKey and AssetCheckKey behave more similarly throughout the product. This PR simply creates a common base class for each of these classes to inherit from.

### CODE REVIEW

**Comment by OwenKephart** [comment]
* **#23630** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23630?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23629** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23629?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by schrockn — CHANGES_REQUESTED** [review]
I'd love to understand the broader context and plan for this. Seems worthy of a github discussion. My initial instinct a superclass isn't necessarily what we want but we can get into it later.

tl;dr I want to understand the whole plan before viewing incremental PRs in a stack.

**Comment by OwenKephart** [comment]
@schrockn https://github.com/dagster-io/internal/discussions/11067
